risingwavelabs/risingwave
# refactor(meta): unify to modify hummock version via transaction

I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Introduce `HummockVersionTransaction` and `HummockVersionStatsTransaction`. 

`HummockVersionTransaction` combines the following:
* apply version delta to latest hummock version
* add version delta to btree-map
* update stats
* notify version delta

`HummockVersionStatsTransaction` combines the following:
* update the in memory hummock version stats
* notify version stats

Previously, the `apply_version_delta` returns `SplitSstInfo`. It is only used to check whether existing compact tasks are expired when we split compaction group. In this PR, we changed to use the `is_compact_task_expired` to do the check, so that `apply_version_delta` won't return the `SplitSstInfo`.

## Checklist

- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.




</details>
